From 6dcd0a4066cd56da2a0b3ca35efceeb5dd97c23b Mon Sep 17 00:00:00 2001
From: Dimitri John Ledkov <dimitri.j.ledkov@intel.com>
Date: Tue, 10 Feb 2015 01:17:09 +0000
Subject: [PATCH 1/2] Add support for ignore_missing attributed in includedir
 bus config.
Organization: Intel Corporation (UK) Ltd. - Co. Reg. #1134945 - Pipers Way, Swindon SN3 1RJ

---
 bus/config-parser.c | 56 +++++++++++++++++++++++++++++++++++++++++++++--------
 doc/busconfig.dtd   |  2 ++
 2 files changed, 50 insertions(+), 8 deletions(-)

diff --git a/bus/config-parser.c b/bus/config-parser.c
index ee2d4e7..513a55c 100644
--- a/bus/config-parser.c
+++ b/bus/config-parser.c
@@ -818,15 +818,39 @@ start_busconfig_child (BusConfigParser   *parser,
     }
   else if (element_type == ELEMENT_INCLUDEDIR)
     {
-      if (!check_no_attributes (parser, "includedir", attribute_names, attribute_values, error))
-        return FALSE;
+      Element *e;
+      const char *ignore_missing;
 
-      if (push_element (parser, ELEMENT_INCLUDEDIR) == NULL)
+      if ((e = push_element (parser, ELEMENT_INCLUDEDIR)) == NULL)
         {
           BUS_SET_OOM (error);
           return FALSE;
         }
 
+      e->d.include.ignore_missing = FALSE;
+
+      if (!locate_attributes (parser, "includedir",
+                              attribute_names,
+                              attribute_values,
+                              error,
+                              "ignore_missing", &ignore_missing,
+                              NULL))
+        return FALSE;
+
+      if (ignore_missing != NULL)
+        {
+          if (strcmp (ignore_missing, "yes") == 0)
+            e->d.include.ignore_missing = TRUE;
+          else if (strcmp (ignore_missing, "no") == 0)
+            e->d.include.ignore_missing = FALSE;
+          else
+            {
+              dbus_set_error (error, DBUS_ERROR_FAILED,
+                              "ignore_missing attribute must have value \"yes\" or \"no\"");
+              return FALSE;
+            }
+        }
+
       return TRUE;
     }
   else if (element_type == ELEMENT_STANDARD_SESSION_SERVICEDIRS)
@@ -2223,6 +2247,7 @@ servicehelper_path (BusConfigParser   *parser,
 static dbus_bool_t
 include_dir (BusConfigParser   *parser,
              const DBusString  *dirname,
+             dbus_bool_t        ignore_missing,
              DBusError         *error)
 {
   DBusString filename;
@@ -2312,6 +2350,7 @@ include_dir (BusConfigParser   *parser,
       goto failed;
     }
 
+ success:
   retval = TRUE;
   
  failed:
@@ -2484,7 +2523,8 @@ bus_config_parser_content (BusConfigParser   *parser,
             goto nomem;
           }
         
-        if (!include_dir (parser, &full_path, error))
+        if (!include_dir (parser, &full_path,
+                          e->d.include.ignore_missing, error))
           {
             _dbus_string_free (&full_path);
             return FALSE;
diff --git a/doc/busconfig.dtd b/doc/busconfig.dtd
index 0cc519b..61491c3 100644
--- a/doc/busconfig.dtd
+++ b/doc/busconfig.dtd
@@ -16,6 +16,8 @@
 <!ELEMENT user (#PCDATA)>
 <!ELEMENT listen (#PCDATA)>
 <!ELEMENT includedir (#PCDATA)>
+<!ATTLIST includedir
+          ignore_missing (yes|no) "no">
 <!ELEMENT servicedir (#PCDATA)>
 <!ELEMENT servicehelper (#PCDATA)>
 <!ELEMENT auth (#PCDATA)>
-- 
2.1.0

